### PR TITLE
add categorical class

### PR DESCRIPTION
**Pull request by pangyoki**
### PR types

Others

### PR changes

APIs

### Describe

- add `Categorical` class
    It's a special case of `multinomial` ( refer to PR https://github.com/PaddlePaddle/Paddle/pull/27219 ). `Categorical` distribution is a discrete probability distribution that describes the possible results of a random variable that can take on one of K possible categories, with the probability of each category separately specified.

- add `sample` method
    Using `multinomial` to sample from `Categorical` distribution.

- add `probs` and `log_prob` method
    Given category index, return the probability of this category.

### CODE REVIEW

File: python/paddle/distribution.py [hunk 1]
```diff
@@ -640,3 +642,226 @@ def kl_divergence(self, other):
...
+
+    * :math:`[x=i]` : it evaluates to 1 if :math:`x==i` , 0 otherwise.
+
+    Args:
+        logits(list|numpy.ndarray|Tensor): The logits input of categorical distribution. The data type is float32 or float64.
+
+    Examples:
+        .. code-block:: python
+
+          import numpy as np
```
Comment (zhiqiu) [inline python/paddle/distribution.py]: Numpy is useless.

File: python/paddle/distribution.py [hunk 2]
```diff
@@ -640,3 +642,226 @@ def kl_divergence(self, other):
...
+          import paddle
+          from paddle.distribution import Categorical
+
+          paddle.disable_static()
```
Comment (zhiqiu) [inline python/paddle/distribution.py]: It can be removed.

File: python/paddle/distribution.py [hunk 3]
```diff
@@ -640,3 +642,226 @@ def kl_divergence(self, other):
...
+
+          x = paddle.rand([6])
+          x.numpy()
+          # [0.32564053, 0.99334985, 0.99034804,
+          #  0.09053693, 0.30820143, 0.19095989]
+          y = paddle.rand([6])
+          y.numpy()
```
Comment (zhiqiu) [inline python/paddle/distribution.py]: `y.numpy()` shows nothing is not if it is not executed in the shell, `print(y,numpy())` is better?

File: python/paddle/distribution.py [hunk 4]
```diff
@@ -640,3 +642,226 @@ def kl_divergence(self, other):
...
+
+            index_prefix = nn.unsqueeze(arange(num_dist), axes=-1)
+            index_prefix = nn.expand(index_prefix, [1, num_value_in_one_dist])
+            index_prefix = nn.unsqueeze(index_prefix, axes=-1)
+
+            index = concat([index_prefix, index_value], axis=-1)
+
+        # value is the category index to search for the corresponding probability.
+        select_prob = gather_nd(prob, index)
+        print(name)
```
Comment (zhiqiu) [inline python/paddle/distribution.py]: Remove debug code.

File: python/paddle/distribution.py [hunk 5]
```diff
@@ -640,3 +642,226 @@ def kl_divergence(self, other):
...
+            index = index_value
+        else:
+            num_dist = np.prod(shape[:-1])
+            num_value_in_one_dist = value_shape[-1]
+            prob = nn.reshape(prob, [num_dist, shape[-1]])
+            if len(value_shape) == 1:
+                value = nn.expand(value, [num_dist])
+                value_shape = shape[:-1] + value_shape
+            index_value = nn.reshape(value, [num_dist, -1, 1])
+            # assert(num_dist == np.prod(value_shape[:-1]), 'dist shape should be the same')
```
Comment (zhiqiu) [inline python/paddle/distribution.py]: Clean the commented code.

File: python/paddle/distribution.py (hunk 1 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: done

File: python/paddle/distribution.py (hunk 2 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: done

File: python/paddle/distribution.py (hunk 3 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: done

File: python/paddle/distribution.py (hunk 4 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: done

File: python/paddle/distribution.py (hunk 5 above)
Comment (pangyoki) [reply python/paddle/distribution.py]: done

File: python/paddle/distribution.py [hunk 6]
```diff
@@ -640,3 +641,318 @@ def kl_divergence(self, other):
...
+    .. math::
+
+        pmf(k; p_i) = \prod_{i=1}^{k} p_i^{[x=i]}
+
+    In the above equation:
+
+    * :math:`[x=i]` : it evaluates to 1 if :math:`x==i` , 0 otherwise.
+
+    Args:
+        logits(list|numpy.ndarray|Tensor): The logits input of categorical distribution. The data type is float32 or float64.
```
Comment (pangyoki) [inline python/paddle/distribution.py:664]: add parameter `name`, will refine it in next PR.

File: python/paddle/distribution.py [hunk 7]
```diff
@@ -640,3 +641,318 @@ def kl_divergence(self, other):
...
+
+          cat.log_prob(value)
+          # [-1.07408 -1.07105 -3.46638]
+
+    """
+
+    def __init__(self, logits, name=None):
+        """
+        Args:
+            logits(list|numpy.ndarray|Variable): The logits input of categorical distribution. The data type is float32 or float64.
```
Comment (pangyoki) [inline python/paddle/distribution.py:706]: `Variable` change to `Tensor`, add parameter `name`, will refine it in next PR.

File: python/paddle/distribution.py [hunk 8]
```diff
@@ -640,3 +641,318 @@ def kl_divergence(self, other):
...
+        return nn.reshape(sample_index, sample_shape, name=name)
+
+    def kl_divergence(self, other):
+        """The KL-divergence between two Categorical distributions.
+
+        Args:
+            other (Categorical): instance of Categorical. The data type is float32.
+
+        Returns:
+            Variable: kl-divergence between two Categorical distributions.
```
Comment (pangyoki) [inline python/paddle/distribution.py:778]: change `Variable` to `Tensor`, will refine it in next PR.

File: python/paddle/distribution.py [hunk 9]
```diff
@@ -640,3 +641,318 @@ def kl_divergence(self, other):
...
+            keep_dim=True,
+            name=name)
+
+        return kl
+
+    def entropy(self):
+        """Shannon entropy in nats.
+
+        Returns:
+          Variable: Shannon entropy of Categorical distribution. The data type is float32.
```
Comment (pangyoki) [inline python/paddle/distribution.py:826]: change `Variable` to `Tensor`, will refine it in next PR.

**Review by zhiqiu — APPROVED** [review]

**Review by kolinwei — APPROVED** [review]

**Review by jzhang533 — APPROVED** [review]
lgtm

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/distribution.py [hunk 10]
```diff
@@ -640,3 +641,318 @@ def kl_divergence(self, other):
...
+        """Generate samples of the specified shape.
+
+        Args:
+          shape (list): Shape of the generated samples.
+
+        Returns:
+          Tensor: A tensor with prepended dimensions shape.
+        
+        Examples:
+        .. code-block:: python
```
Comment (pangyoki) [inline python/paddle/distribution.py:736]: ``code-block:: python` needs indentation. The same below.

File: python/paddle/distribution.py (hunk 6 above)
Comment (pangyoki) [reply python/paddle/distribution.py:664]: done, refer to PR #27946

File: python/paddle/distribution.py (hunk 7 above)
Comment (pangyoki) [reply python/paddle/distribution.py:706]: done, refer to PR #27946

File: python/paddle/distribution.py (hunk 8 above)
Comment (pangyoki) [reply python/paddle/distribution.py:778]: done, refer to PR #27946

File: python/paddle/distribution.py (hunk 9 above)
Comment (pangyoki) [reply python/paddle/distribution.py:826]: done, refer to PR #27946

File: python/paddle/distribution.py (hunk 10 above)
Comment (pangyoki) [reply python/paddle/distribution.py:736]: done, refer to PR #27946

**Pull request merged**